oauth2_http/java/com/google/auth/oauth2/AwsRequestSignature.java

```diff
@@ -0,0 +1,193 @@
+      List<String> sortedHeaderNames,
+      String signature,
+      String credentialScope,
+      String url,
+      String httpMethod,
+      String date,
+      String region) {
+    this.awsSecurityCredentials = awsSecurityCredentials;
+    this.canonicalHeaders = canonicalHeaders;
+    this.sortedHeaderNames = sortedHeaderNames;
+    this.signature = signature;
+    this.credentialScope = credentialScope;
+    this.url = url;
+    this.httpMethod = httpMethod;
+    this.date = date;
+    this.region = region;
+  }
+
+  /** Returns the request signature based on the AWS Signature Version 4 signing process. */
+  String getSignature() {
+    return signature;
+  }
+
+  /** Returns the credential scope. e.g. 20150830/us-east-1/iam/aws4_request */
+  String getCredentialScope() {
+    return credentialScope;
+  }
+
+  /** Returns the AWS security credentials. */
+  AwsSecurityCredentials getSecurityCredentials() {
+    return awsSecurityCredentials;
+  }
+
+  /** Returns the request URL. */
+  String getUrl() {
+    return url;
+  }
+
+  /** Returns the HTTP request method. */
+  String getHttpMethod() {
+    return httpMethod;
+  }
+
+  /** Returns the HTTP request canonical headers. */
+  Map<String, String> getCanonicalHeaders() {
+    return new HashMap<>(canonicalHeaders);
+  }
+
+  /** Returns the request date. */
+  String getDate() {
+    return date;
+  }
+
+  /** Returns a list of headers sorted in lexicographical order. */
+  List<String> getSortedHeaderNames() {
+    return new ArrayList<>(sortedHeaderNames);
+  }
+
+  /** Returns the targeted region. */
+  String getRegion() {
+    return region;
+  }
+
+  static class Builder {
+
+    private AwsSecurityCredentials awsSecurityCredentials;
+    private Map<String, String> canonicalHeaders;
+    private List<String> sortedHeaderNames;
+
+    private String signature;
+    private String credentialScope;
+    private String url;
+    private String httpMethod;
+    private String date;
+    private String region;
+
+    Builder setSignature(String signature) {
+      this.signature = signature;
+      return this;
+    }
+
+    Builder setCredentialScope(String credentialScope) {
+      this.credentialScope = credentialScope;
+      return this;
+    }
+
+    Builder setSecurityCredentials(AwsSecurityCredentials awsSecurityCredentials) {
+      this.awsSecurityCredentials = awsSecurityCredentials;
+      return this;
+    }
+
+    Builder setUrl(String url) {
+      this.url = url;
+      return this;
+    }
+
+    Builder setHttpMethod(String httpMethod) {
+      this.httpMethod = httpMethod;
+      return this;
+    }
+
+    Builder setCanonicalHeaders(Map<String, String> canonicalHeaders) {
+      this.canonicalHeaders = new HashMap<>(canonicalHeaders);
+      return this;
+    }
+
+    Builder setDate(String date) {
+      this.date = date;
+      return this;
+    }
+
+    Builder setSortedHeaderNames(List<String> sortedHeaderNames) {
```

Comment:
what if these come in unsorted? Nothing here sorts them or checks that they're sorted.

Reply:
Usuability wise, it might be easier for this to not be sorted and to ensure the sorting when we do the signing. Otherwise, it'll be quite a few places where we try to ensure sorting.

Reply:
Good points. I shouldn't have added this - removed.